izpack-core/src/main/java/com/izforge/izpack/core/data/DefaultVariables.java

```diff
@@ -294,89 +296,133 @@ public synchronized void add(DynamicVariable variable)
         dynamicVariables.add(variable);
     }
 
-    private String evaluateVariable(DynamicVariable variable)
-    {
-        String newValue;
-        try
-        {
-            newValue = variable.evaluate(replacer);
-        }
-        catch (IzPackException exception)
-        {
-            throw exception;
-        }
-        catch (Exception exception)
-        {
-            throw new IzPackException("Failed to refresh dynamic variable (" + variable.getName() + ")", exception);
-        }
-        return newValue;
-    }
-
     /**
      * Refreshes dynamic variables.
-     * This method relies on a pre-ordered list @{code dynamicVariables} according to referred variables to
-     * resolve them in the right order.
      *
      * @throws InstallerException if variables cannot be refreshed
      */
     @Override
     public synchronized void refresh() throws InstallerException
     {
         logger.fine("Refreshing dynamic variables");
-
+        Set<DynamicVariable> checkedVariables = new HashSet<DynamicVariable>();
+        Set<String> unsetVariables = new HashSet<String>();
         Set<String> setVariables = new HashSet<String>();
+        // for dependent dynamic variables a size of dynamicVariables.size()+1 would be enough
+        // in case of conditions, which depend on dynamic variables also, we need more iterations
+        // to be on the safe side, we take 10*dynamicVariables.size()+1
+        int maxCount = 10*dynamicVariables.size()+1;
+        int count=maxCount;
+        boolean changed = true;
+        while (changed) {
```

Comment:
I'm still doing tests, but not to block you from further changes - I removed the loop in commit f6b1a5688c5de5638064fde06b9d74317ba811e2.